lib/cpp/src/thrift/protocol/TCompactProtocol.tcc

```diff
@@ -387,7 +389,11 @@ uint32_t TCompactProtocolT<Transport_>::writeVarint64(uint64_t n) {
  */
 template <class Transport_>
 uint64_t TCompactProtocolT<Transport_>::i64ToZigzag(const int64_t l) {
-  return (l << 1) ^ (l >> 63);
+  const int64_t m = l >> 63;
+  uint64_t p, q;
+  memcpy(&p, &l, sizeof(p));
+  memcpy(&q, &m, sizeof(q));
+  return p = (p << 1) ^ q;
```

Comment:
`return (static_cast<uint64_t>(l) << 1) ^ (l >> 63);` would be much simpler and efficient.
static casting of negative to unsigned value is well defined AFAICT.